### DISCUSSION TITLE
View-Mode: Support WebView Download capabilities and saving to file

### DISCUSSION THREAD

**Discussion by xi7** [gsantner/markor#2314] [Q&A] [answered]
![Screenshot_20240531_103212_Markor_1](https://github.com/gsantner/markor/assets/67331785/b471ed21-d8b2-4503-8887-5cc15bddf132)

after press save data using viewmode, where i find this file ?

**Comment by gsantner** [comment]
In the top toolbar menu you can find a info option. You can find the folder info there.

Also you should be able to see it in the file manager. Where you saved the file you usually know as you create/select it.

**Reply by xi7** [reply]
> In the top toolbar menu you can find a info option. You can find the folder info there. Also you should be able to see it in the file manager. Where you saved the file you usually know as you create/select it.

no, is not this... i mean my table data after save where i find my file json. 

save and load working... but i cant find it

**Reply by gsantner** [reply]
I dont know? You create the file and edit it's contents. Saving writes the same file again.

If it's about View mode, on it's own doesn't have edit/save fratures. Just to view file's contents.

I don't know about this table in screenshot. Your content, your choice of file hierachy and structure.

**Reply by xi7** [reply]
> I dont know? You create the file and edit it's contents. Saving writes the same file again.
> 
> If it's about View mode, on it's own doesn't have edit/save fratures. Just to view file's contents.
> 
> I don't know about this table in screenshot. Your content, your choice of file hierachy and structure.


https://github.com/gsantner/markor/assets/67331785/0918c6f5-22cb-44fd-9e76-73cc1e662fef

check video if u understand, after press import data will load correctly using viewmode, but where is the file ?

**Reply by gsantner** [reply]
Still, this is your own stuff. If you didn't write this html file you need to ask at whom have written it how it works.

This javascript save popup is not related to Markor file management features, it's some custom popup in your html/javascript. That "Successful save saying" is not a saying from Markor.

**Reply by xi7** [reply]
> Still, this is your own stuff. If you didn't write this html file you need to ask at whom have written it how it works.
> 
> This javascript save popup is not related to Markor file management features, it's some custom popup in your html/javascript. That "Successful save saying" is not a saying from Markor.

this is my file...
In summary, saveData saves player data to local storage, and loadData retrieves and populates player data from local storage when the page is loaded or refreshed.

it stores the data array as a JSON string in the browser's local storage under the key 'playerData'.

but where is my data file :(

**Reply by gsantner** [reply]
localStorage is in browser data. You won't find it anywhere as dedicated file in a file browser. But you should be able to read/write it later by same javascript APIs.

**Reply by xi7** [reply]
> localStorage está nos dados do navegador. Você não vai encontrá-lo em qualquer lugar como arquivo dedicado em um navegador de arquivos. Mas você deve ser capaz de lê-lo / gravá-lo mais tarde pelas mesmas APIs javascript.

im trying just backup data using viewmode, it is impossible there ?

**Reply by gsantner** [reply]
You can do not (much) more or less what you can do with normal android web browser. The web doesn't have direct  read/write file access. Thats a important security feature.

Basically what happens in browser (view mode) stays in view mode. You could i.e. copy stuff to clipboard to update/edit a file.

You are doing web development here, not a Markor specific restriction/issue. You might get some ideas on the possibilities in web development sites.

**Reply by xi7** [reply]
> You can do not (much) more or less what you can do with normal android web browser. The web doesn't have direct read/write file access. Thats a important security feature.
> 
> Basically what happens in browser (view mode) stays in view mode. You could i.e. copy stuff to clipboard to update/edit a file.
> 
> You are doing web development here, not a Markor specific restriction/issue. You might get some ideas on the possibilities in web development sites.

![Screenshot_20240531_160036_Firefox](https://github.com/gsantner/markor/assets/67331785/fc1036d8-71d2-40c2-a111-214fac96ec1b)

using firefox android the import/export save to file json and load all data, but markor viewmode the button no work there.



also HTML editor app, working normally with viewmode.
![Screenshot_20240531_165052_Html Editor_1](https://github.com/gsantner/markor/assets/67331785/6fe11766-cdff-494e-9a3a-3d1190d6b1af)

**Comment by harshad1** [comment]
@gsantner we do have this:

https://github.com/gsantner/markor/blob/master/app/src/main/java/net/gsantner/markor/activity/DocumentEditAndViewFragment.java#L164

It has been in markor since I remember. Should we disable this for safety?

**Comment by harshad1** [comment]
Fundamentally, it seems that JS in firefox can trigger a download, but not in markor's webview?

**Reply by xi7** [reply]
> Fundamentally, it seems that JS in firefox can trigger a download, but not in markor's webview?

yes, this work with all browser... and apps like trebedit, html editor, but not work using markor view

**Comment by gsantner** [comment, marked as answer]
-> Issue #2315 

OK to sum up these loads of text: You have a feature request to support file downloads in Markor View Mode, and to chose where to save that file. If you initially would attached these screenshots instead of asking where files are saved and video of local-file webapp...would be much clearer.

So ... file downloads are currently not handled. Nobody came up with the idea to use Markor in that way yet...for more sophisticated webapps with downloads 😄 .

-> Thanks for the feature request and feel free to implement this and make a merge request to get the change in for all Markor users.
